netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -73,7 +71,34 @@ private ProtocolNegotiators() {
   public static ChannelHandler serverTls(SSLEngine sslEngine) {
     Preconditions.checkNotNull(sslEngine, "sslEngine");
 
-    return new SslHandler(sslEngine, false);
+    final SslHandler sslHandler = new SslHandler(sslEngine, false);
+    return new AbstractNegotiationHandler() {
+      @Override
+      public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        super.handlerAdded(ctx);
+        ctx.pipeline().addFirst(sslHandler);
+      }
+
+      @Override
+      public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        if (evt instanceof SslHandshakeCompletionEvent) {
+          SslHandshakeCompletionEvent handshakeEvent = (SslHandshakeCompletionEvent) evt;
+          if (handshakeEvent.isSuccess()) {
+            SslHandler handler = ctx.pipeline().get(SslHandler.class);
+            if (handler.applicationProtocol() != null) {
+              // Successfully negotiated the protocol.
+              ctx.pipeline().remove(this);
```

Comment:
Use `writeBufferedAndRemove()` or don't use `AbstractNegotiationHandler`. I don't think anything will notice the exceptions passed to failed, fyi.